File: services/scada-gate-control/src/domain/machine-boundary.ts
/**
 * TypeScript mirror of the shared machine-boundary v1 contracts (PR 6.0).
 *
 * These readonly types mirror `contracts/machine-boundary/v1/*.schema.json`. The
 * `*_FIELDS` tuples and enum tuples are asserted against the schema by the
 * contract spec so a drift fails a test. Type/library surface only in PR 6.0: no
 * route, CommandService, GateController, transport, or audit code imports it. 6.2
 * consumes the command-intent + emits the receipt; 6.1a produces the snapshot.
 */
import type { Brand } from '../types';

export type DeviceId = Brand<string, 'DeviceId'>;
export type AdapterGateId = Brand<string, 'AdapterGateId'>;
export type CanonicalGateId = Brand<string, 'CanonicalGateId'>;
export type IntentId = Brand<string, 'IntentId'>;
export type CorrelationId = Brand<string, 'CorrelationId'>;
export type ReceiptId = Brand<string, 'ReceiptId'>;
export type CampaignId = Brand<string, 'CampaignId'>;
export type PlanId = Brand<string, 'PlanId'>;
export type RequirementRunId = Brand<string, 'RequirementRunId'>;
export type ModelSnapshotId = Brand<string, 'ModelSnapshotId'>;
export type PredictionRunId = Brand<string, 'PredictionRunId'>;
export type CapabilityReleaseId = Brand<string, 'CapabilityReleaseId'>;
export type Sha256 = Brand<string, 'Sha256'>;
export type UtcInstant = Brand<string, 'UtcInstant'>;
export type RequestId = Brand<string, 'RequestId'>;
export type IdempotencyKey = Brand<string, 'IdempotencyKey'>;

export const MACHINE_BOUNDARY_SCHEMA_VERSION = 1 as const;
// Must equal contracts/machine-boundary/v1/manifest.json:contract_set_sha256 and
// the Scheduler MACHINE_BOUNDARY_CONTRACT_SET_SHA256 constant.
export const MACHINE_BOUNDARY_CONTRACT_SET_SHA256 =
  '155606e21b0e15dc674a4fb3e5a9e2f40b94d6e1078ae09c973944731f3db0e4';

export const EVENT_KINDS = ['open', 'trim', 'close'] as const;
export type EventKind = (typeof EVENT_KINDS)[number];

export const COMMAND_MODES = ['shadow', 'operator_approved'] as const;
export type CommandMode = (typeof COMMAND_MODES)[number];

export const VALIDATION_STATUSES = ['validation_accepted', 'validation_rejected'] as const;
export type ValidationStatus = (typeof VALIDATION_STATUSES)[number];

export const VALIDATION_REJECTION_REASONS = [
  'schema_invalid',
  'capability_mismatch',
  'target_invalid',
  'not_before_violation',
  'deadline_expired',
  'lineage_mismatch',
  'freshness_failed',
  'idempotency_conflict',
] as const;
export type ValidationRejectionReason = (typeof VALIDATION_REJECTION_REASONS)[number];

export type DeviceTargetCapability = {
  readonly target_position_m: number;
  readonly target_level: number;
};

export type DeviceCapability = {
  readonly device_id: DeviceId;
  readonly adapter_gate_id: AdapterGateId;
  readonly targets: readonly DeviceTargetCapability[];
};

export type DeviceCapabilitySnapshot = {
  readonly schema_version: 1;
  readonly capability_release_id: CapabilityReleaseId;
  readonly capability_hash: Sha256;
  // Keyed by canonical_gate_id -> capability: one gate maps to at most one device.
  readonly capabilities: Readonly<Record<CanonicalGateId, DeviceCapability>>;
};

export type CommandLineage = {
  readonly campaign_id: CampaignId;
  readonly plan_id: PlanId;
  readonly plan_version: number;
  readonly input_content_hash: Sha256;
  readonly draft_content_hash: Sha256;
  readonly requirement_run_id: RequirementRunId;
  readonly requirement_version: number;
  readonly requirement_set_sha256: Sha256;
  readonly model_snapshot_id: ModelSnapshotId;
  readonly model_release_id: string;
  readonly model_release_content_hash: Sha256;
  readonly prediction_run_id: PredictionRunId;
  readonly prediction_identity_version: 2;
  readonly engine_descriptor_content_hash: Sha256;
  readonly artifact_sha256: Sha256;
};

export type CommandIntent = {
  readonly schema_version: 1;
  readonly intent_id: IntentId;
  readonly correlation_id: CorrelationId;
  readonly request_id: RequestId;
  readonly idempotency_key: IdempotencyKey;
  readonly canonical_gate_id: CanonicalGateId;
  readonly event_kind: EventKind;
  readonly event_sequence: number;
  readonly gate_event_sequence: number;
  readonly device_id: DeviceId;
  readonly adapter_gate_id: AdapterGateId;
  readonly capability_release_id: CapabilityReleaseId;
  readonly capability_hash: Sha256;
  readonly target_position_m: number;
  readonly target_level: number;
  readonly not_before: UtcInstant;
  readonly deadline: UtcInstant;
  readonly mode: CommandMode;
  readonly lineage: CommandLineage;
};

export type ValidationReceipt = {
  readonly schema_version: 1;
  readonly receipt_id: ReceiptId;
  readonly intent_id: IntentId;
  readonly correlation_id: CorrelationId;
  readonly request_id: RequestId;
  readonly idempotency_key: IdempotencyKey;
  readonly intent_content_hash: Sha256;
  readonly capability_hash: Sha256;
  readonly status: ValidationStatus;
  readonly validated_at: UtcInstant;
  readonly reason_code: ValidationRejectionReason | null;
};

// Field-key rosters asserted against the schema properties by the contract spec.
export const COMMAND_INTENT_FIELDS = [
  'schema_version',
  'intent_id',
  'correlation_id',
  'request_id',
  'idempotency_key',
  'canonical_gate_id',
  'event_kind',
  'event_sequence',
  'gate_event_sequence',
  'device_id',
  'adapter_gate_id',
  'capability_release_id',
  'capability_hash',
  'target_position_m',
  'target_level',
  'not_before',
  'deadline',
  'mode',
  'lineage',
] as const;

export const COMMAND_LINEAGE_FIELDS = [
  'campaign_id',
  'plan_id',
  'plan_version',
  'input_content_hash',
  'draft_content_hash',
  'requirement_run_id',
  'requirement_version',
  'requirement_set_sha256',
  'model_snapshot_id',
  'model_release_id',
  'model_release_content_hash',
  'prediction_run_id',
  'prediction_identity_version',
  'engine_descriptor_content_hash',
  'artifact_sha256',
] as const;

export const VALIDATION_RECEIPT_FIELDS = [
  'schema_version',
  'receipt_id',
  'intent_id',
  'correlation_id',
  'request_id',
  'idempotency_key',
  'intent_content_hash',
  'capability_hash',
  'status',
  'validated_at',
  'reason_code',
] as const;

export const DEVICE_CAPABILITY_SNAPSHOT_FIELDS = [
  'schema_version',
  'capability_release_id',
  'capability_hash',
  'capabilities',
] as const;

export const DEVICE_CAPABILITY_FIELDS = ['device_id', 'adapter_gate_id', 'targets'] as const;

export const DEVICE_TARGET_FIELDS = ['target_position_m', 'target_level'] as const;

File: services/scheduler/src/schemas/machine_boundary.py
"""Strict Python mirrors of the shared machine-boundary v1 contracts (PR 6.0).

These pydantic models mirror `contracts/machine-boundary/v1/*.schema.json` field
for field so a drift between the JSON Schema (validated by jsonschema here and by
Ajv in the TypeScript SCADA service) and the Python model fails a test. TEST /
library surface only in PR 6.0: NOT imported by main.py, routes, repositories, or
schemas/__init__.py; emits no intents. 4.3c compiles intents into `CommandIntent`;
6.2 consumes it + produces `ValidationReceipt`; 6.1a produces
`DeviceCapabilitySnapshot`.

Cross-language parity: identifiers are printable ASCII (`[!-~]+`, no Unicode
`\\S`), numbers are finitely bounded, integers use mathematical-integer semantics
(a whole-number float is accepted), and timestamps enforce field ranges — so
Python and Ajv agree on every input, not just the fixtures. Deadline ordering,
capability freshness, quantizer membership, and full calendar validation (leap
day) are runtime semantics owned by 4.3c/6.1a/6.2.
"""

from __future__ import annotations

import math
from typing import Annotated, Any, Dict, List, Literal, Optional

from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

MACHINE_BOUNDARY_SCHEMA_VERSION = 1
# Must equal contracts/machine-boundary/v1/manifest.json:contract_set_sha256. Both
# this module and the TypeScript SCADA mirror pin it so a schema/fixture/expected
# change without a coordinated version bump fails both suites.
MACHINE_BOUNDARY_CONTRACT_SET_SHA256 = (
    "155606e21b0e15dc674a4fb3e5a9e2f40b94d6e1078ae09c973944731f3db0e4"
)


def _strict_int(value: Any) -> Any:
    # Mathematical-integer semantics matching JSON Schema `type: integer`: accept a
    # JSON integer or a whole-number float (1.0 -> 1); reject bool, str, and a
    # non-integral float (3.5).
    if isinstance(value, bool):
        raise ValueError("expected a JSON integer")
    if isinstance(value, int):
        return value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    raise ValueError("expected a JSON integer")


def _strict_number(value: Any) -> Any:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError("expected a JSON number")
    if isinstance(value, float) and not math.isfinite(value):
        raise ValueError("number must be finite")
    return value


_UUID_RE = (
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)
_SHA256_RE = r"^[0-9a-f]{64}$"
_REQUEST_ID_RE = r"^[A-Za-z0-9._-]{1,128}$"
_ID_TOKEN_RE = r"^[!-~]+$"  # printable ASCII, engine-independent (no Unicode \S)
_UTC_RE = (
    r"^[0-9]{4}-(0[1-9]|1[0-2])-(0[1-9]|[12][0-9]|3[01])"
    r"T(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9](\.[0-9]{1,6})?Z$"
)

Uuid = Annotated[str, StringConstraints(min_length=36, max_length=36, pattern=_UUID_RE)]
Sha256 = Annotated[
    str, StringConstraints(min_length=64, max_length=64, pattern=_SHA256_RE)
]
RequestId = Annotated[str, StringConstraints(pattern=_REQUEST_ID_RE)]
IdToken = Annotated[
    str, StringConstraints(min_length=1, max_length=128, pattern=_ID_TOKEN_RE)
]
ReleaseId = Annotated[
    str, StringConstraints(min_length=1, max_length=256, pattern=_ID_TOKEN_RE)
]
UtcInstant = Annotated[
    str, StringConstraints(min_length=20, max_length=27, pattern=_UTC_RE)
]
PositiveInt = Annotated[int, BeforeValidator(_strict_int), Field(ge=1)]
TargetLevel = Annotated[int, BeforeValidator(_strict_int), Field(ge=0, le=65535)]
PositionM = Annotated[float, BeforeValidator(_strict_number), Field(ge=0, le=1000)]

EventKind = Literal["open", "trim", "close"]
CommandMode = Literal["shadow", "operator_approved"]
ValidationStatus = Literal["validation_accepted", "validation_rejected"]
ValidationRejectionReason = Literal[
    "schema_invalid",
    "capability_mismatch",
    "target_invalid",
    "not_before_violation",
    "deadline_expired",
    "lineage_mismatch",
    "freshness_failed",
    "idempotency_conflict",
]


class _MachineBoundaryModel(BaseModel):
    model_config = ConfigDict(extra="forbid", protected_namespaces=())


class DeviceTargetCapability(_MachineBoundaryModel):
    target_position_m: PositionM
    target_level: TargetLevel


class DeviceCapability(_MachineBoundaryModel):
    device_id: IdToken
    adapter_gate_id: IdToken
    targets: Annotated[List[DeviceTargetCapability], Field(min_length=1)]

    @field_validator("targets")
    @classmethod
    def _targets_unique(cls, value: List[DeviceTargetCapability]):
        seen = set()
        for target in value:
            key = (target.target_position_m, target.target_level)
            if key in seen:
                raise ValueError("targets must be unique (schema uniqueItems)")
            seen.add(key)
        return value


class DeviceCapabilitySnapshot(_MachineBoundaryModel):
    schema_version: Literal[1]
    capability_release_id: IdToken
    capability_hash: Sha256
    # A map keyed by canonical_gate_id -> capability: one gate maps to at most one
    # device (structural uniqueness). Empty map is the valid dark default.
    capabilities: Dict[IdToken, DeviceCapability]


class CommandLineage(_MachineBoundaryModel):
    campaign_id: Uuid
    plan_id: Uuid
    plan_version: PositiveInt
    input_content_hash: Sha256
    draft_content_hash: Sha256
    requirement_run_id: Uuid
    requirement_version: PositiveInt
    requirement_set_sha256: Sha256
    model_snapshot_id: Sha256
    model_release_id: ReleaseId
    model_release_content_hash: Sha256
    prediction_run_id: Sha256
    prediction_identity_version: Literal[2]
    engine_descriptor_content_hash: Sha256
    artifact_sha256: Sha256


class CommandIntent(_MachineBoundaryModel):
    schema_version: Literal[1]
    intent_id: Uuid
    correlation_id: Uuid
    request_id: RequestId
    idempotency_key: IdToken
    canonical_gate_id: IdToken
    event_kind: EventKind
    event_sequence: PositiveInt
    gate_event_sequence: PositiveInt
    device_id: IdToken
    adapter_gate_id: IdToken
    capability_release_id: IdToken
    capability_hash: Sha256
    target_position_m: PositionM
    target_level: TargetLevel
    not_before: UtcInstant
    deadline: UtcInstant
    mode: CommandMode
    lineage: CommandLineage

    @model_validator(mode="after")
    def _event_position_invariant(self) -> "CommandIntent":
        # Mirror gate_plan_events: close -> position 0; open/trim -> position > 0.
        if self.event_kind == "close" and self.target_position_m != 0:
            raise ValueError("a close intent requires target_position_m == 0")
        if self.event_kind in ("open", "trim") and not self.target_position_m > 0:
            raise ValueError("an open/trim intent requires target_position_m > 0")
        return self


class ValidationReceipt(_MachineBoundaryModel):
    schema_version: Literal[1]
    receipt_id: Uuid
    intent_id: Uuid
    correlation_id: Uuid
    request_id: RequestId
    idempotency_key: IdToken
    intent_content_hash: Sha256
    capability_hash: Sha256
    status: ValidationStatus
    validated_at: UtcInstant
    reason_code: Optional[ValidationRejectionReason]

    @model_validator(mode="after")
    def _reason_matches_status(self) -> "ValidationReceipt":
        if self.status == "validation_accepted" and self.reason_code is not None:
            raise ValueError("an accepted receipt must carry a null reason_code")
        if self.status == "validation_rejected" and self.reason_code is None:
            raise ValueError("a rejected receipt must carry a reason_code")
        return self


# manifest schema filename -> mirror model (the contract test binds these).
MODEL_BY_SCHEMA = {
    "command-intent.schema.json": CommandIntent,
    "validation-receipt.schema.json": ValidationReceipt,
    "device-capability-snapshot.schema.json": DeviceCapabilitySnapshot,
}
